packages/react-router/lib/router/router.ts

```diff
@@ -878,69 +879,73 @@ export function createRouter(init: RouterInit): Router {
       pathname: init.history.location.pathname,
     });
     let { matches, route } = getShortCircuitMatches(dataRoutes);
+    initialized = true;
     initialMatches = matches;
     initialErrors = { [route.id]: error };
-  }
-
-  // In SPA apps, if the user provided a patchRoutesOnNavigation implementation and
-  // our initial match is a splat route, clear them out so we run through lazy
-  // discovery on hydration in case there's a more accurate lazy route match.
-  // In SSR apps (with `hydrationData`), we expect that the server will send
-  // up the proper matched routes so we don't want to run lazy discovery on
-  // initial hydration and want to hydrate into the splat route.
-  if (initialMatches && !init.hydrationData) {
-    let fogOfWar = checkFogOfWar(
-      initialMatches,
-      dataRoutes,
-      init.history.location.pathname
-    );
-    if (fogOfWar.active) {
-      initialMatches = null;
+  } else {
```

Comment:
All this stuff after this initial 404 check is for checking fog of war and checking if we need to kick off any lazy/loader functions etc.  But if this was an initial 404 and there is no fog of war implementation, we don't want to bother with any of that since we know we can just initialize with errors synchronously.